realsense_camera/src/base_nodelet.cpp

```diff
@@ -724,81 +726,94 @@ namespace realsense_camera
-      if (depth_scale_meters_ == MILLIMETER_METERS)
+      image_depth16_ = reinterpret_cast<const uint16_t *>(frame.get_data());
+      float depth_scale_meters = rs_get_device_depth_scale(rs_device_, &rs_error_);
+      if (depth_scale_meters == MILLIMETER_METERS)
       {
-        image_[stream_index].data = (unsigned char *) (rs_get_frame_data(rs_device_, stream_index, 0));
+        image_[stream_index].data = (unsigned char *) image_depth16_;               
       }
       else
       {
         cvWrapper_ = cv::Mat(image_[stream_index].size(), cv_type_[stream_index], (void *) image_depth16_,
               step_[stream_index]);
         cvWrapper_.convertTo(image_[stream_index], cv_type_[stream_index],
-              static_cast<double>(depth_scale_meters_) / static_cast<double>(MILLIMETER_METERS));
+              static_cast<double>(depth_scale_meters) / static_cast<double>(MILLIMETER_METERS));
       }
     }
     else
     {
-      image_[stream_index].data = (unsigned char *) (rs_get_frame_data(rs_device_, stream_index, 0));
+      image_[stream_index].data = (unsigned char *) (frame.get_data());
     }
   }
 
   /*
-   * Prepare and publish topics.
+   * Set depth enable
    */
-  void BaseNodelet::prepareTopics() try
+  void BaseNodelet::setDepthEnable(bool &enable_depth)
   {
-    while (ros::ok())
+    // Set flags
+    if (enable_depth == false)
     {
-      if (start_stop_srv_called_ == true)
+      if (enable_[RS_STREAM_COLOR] == false)
       {
-        if (start_camera_ == true)
-        {
-          ROS_INFO_STREAM(nodelet_name_ << " - " << startCamera());
-        }
-        else
-        {
-          ROS_INFO_STREAM(nodelet_name_ << " - " << stopCamera());
-        }
-        start_stop_srv_called_ = false;
+        ROS_INFO_STREAM(nodelet_name_ << " - Color stream is also disabled. Cannot disable depth stream");
+        enable_depth = true;
       }
-
-      if (enable_[RS_STREAM_DEPTH] != rs_is_stream_enabled(rs_device_, RS_STREAM_DEPTH, 0))
+      else
       {
-        stopCamera();
-        setStreams();
-        startCamera();
+        enable_[RS_STREAM_DEPTH] = false;
       }
+    }
+    else
+    {
+      enable_[RS_STREAM_DEPTH] = true;
+    }
 
-      if (rs_is_device_streaming(rs_device_, 0) == 1)
-      {
-        rs_wait_for_frames(rs_device_, &rs_error_);
-        checkError();
+    if (enable_[RS_STREAM_DEPTH] != rs_is_stream_enabled(rs_device_, RS_STREAM_DEPTH, 0))
+    {
+      stopCamera();
+      setStreams();
+      startCamera();
+    }
+  }
 
-        duplicate_depth_color_ = false;
+  /*
+   * Publish topic.
+   */
+  void BaseNodelet::publishTopic(rs_stream stream_index, rs::frame &frame) try
+  {
+    // mutex to ensure only one frame per stream is processed at a time
+    boost::mutex::scoped_lock lock(frame_mutex_[stream_index]);
 
-        publishTopics();
+    if (rs_is_stream_enabled(rs_device_, (rs_stream) stream_index, 0) == 1)
```

Comment:
Since we are called by the callback, we have to be enabled. We should remove this check.